File: Tree/C++/serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string serialized = "";
        recursiveSerialize(root, serialized);
        return serialized;
    }
    
    void recursiveSerialize(TreeNode* root, string &serialized) {
        if (root == NULL) {
            serialized += "null,";
        } else {
            serialized += to_string(root->val) + ",";
            recursiveSerialize(root->left, serialized);
            recursiveSerialize(root->right, serialized);
        }
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        queue<string> vals;
        string s;
        for (int i = 0; i < data.size(); i++) {
            if (data[i] == ',') {
                vals.push(s);
                s = "";
            } else {
                s += data[i];
            }
        }
        return recursiveDeserialize(vals);
    }
    
    TreeNode* recursiveDeserialize(queue<string>& vals) {
        TreeNode* root = new TreeNode(0);
        if (vals.empty()) {
            return NULL;
        }
        if (vals.front() == "null") {
            vals.pop();
            return NULL;
        }
        root->val = stoi(vals.front());
        vals.pop();
        root->left = recursiveDeserialize(vals);
        root->right = recursiveDeserialize(vals);
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: Amazon/Golang/serialize-and-deserialize-binary-tree.go
import "strconv"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Codec struct {
    sb strings.Builder
}

func Constructor() Codec {
    return Codec{strings.Builder{}}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
    this.sb.Reset()
    serializeHelper(root, &this.sb)
    return this.sb.String()
}

func serializeHelper(root *TreeNode, sb *strings.Builder) {
    if root == nil {
        sb.WriteString("null,")
    } else {
        sb.WriteString(strconv.Itoa(root.Val)+",")
        serializeHelper(root.Left, sb)
        serializeHelper(root.Right, sb)
    }
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {    
    nodeStrs := strings.Split(data, ",")
    return deserializeHelper(&nodeStrs)
}

func deserializeHelper(nodeStrs *[]string) *TreeNode {
    if len(*nodeStrs) == 0 {
        return nil
    }
    if (*nodeStrs)[0] == "null" {
        (*nodeStrs) = (*nodeStrs)[1:]
        return nil
    }
    val, _ := strconv.Atoi((*nodeStrs)[0])
    (*nodeStrs) = (*nodeStrs)[1:]
    root := &TreeNode{val, nil, nil}
    root.Left = deserializeHelper(nodeStrs)
    root.Right = deserializeHelper(nodeStrs)
    return root
}


/**
 * Your Codec object will be instantiated and called as such:
 * obj := Constructor();
 * data := obj.serialize(root);
 * ans := obj.deserialize(data);
 */
